File: source/includes/write/bulk-write.java
import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.ServerApi;
import com.mongodb.ServerApiVersion;

import com.mongodb.bulk.BulkWriteResult;
import com.mongodb.client.model.InsertOneModel;
import com.mongodb.client.model.UpdateOneModel;
import com.mongodb.client.model.UpdateManyModel;
import com.mongodb.client.model.ReplaceOneModel;
import com.mongodb.client.model.DeleteOneModel;
import com.mongodb.client.model.DeleteManyModel;
import com.mongodb.client.model.BulkWriteOptions;
import com.mongodb.reactivestreams.client.MongoCollection;

import org.bson.Document;

import com.mongodb.reactivestreams.client.MongoClient;
import com.mongodb.reactivestreams.client.MongoClients;
import com.mongodb.reactivestreams.client.MongoDatabase;
import org.reactivestreams.Publisher;
import reactor.core.publisher.Mono;

import java.util.Arrays;

import static com.mongodb.client.model.Filters.eq;
import static com.mongodb.client.model.Updates.set;

class BulkWrite {
    public static void main(String[] args) throws InterruptedException {
        // Replace the placeholder with your Atlas connection string
        String uri = "<connection string>";

        // Construct a ServerApi instance using the ServerApi.builder() method
        ServerApi serverApi = ServerApi.builder()
                .version(ServerApiVersion.V1)
                .build();

        MongoClientSettings settings = MongoClientSettings.builder()
                .applyConnectionString(new ConnectionString(uri))
                .serverApi(serverApi)
                .build();

        // Create a new client and connect to the server
        try (MongoClient mongoClient = MongoClients.create(settings)) {
            MongoDatabase sample_restaurants = mongoClient.getDatabase("sample_restaurants");
            MongoCollection<Document> restaurants = sample_restaurants.getCollection("restaurants");


            // start-bulk-insert-one
            InsertOneModel<Document> operation = new InsertOneModel<>(
                    new Document("name", "Mongo's Deli")
                            .append("cuisine", "Sandwiches"));
            // end-bulk-insert-one

            // start-bulk-update-one
            UpdateOneModel<Document> operation = new UpdateOneModel<>(
                    eq("name", "Mongo's Deli"),
                    set("cuisine", "Sandwiches and Salads"));
            // end-bulk-update-one

            // start-bulk-update-many
            UpdateManyModel<Document> operation = new UpdateManyModel<>(
                    eq("name", "Mongo's Deli"),
                    set("cuisine", "Sandwiches and Salads"));
            // end-bulk-update-many

            // start-bulk-replace-one
            ReplaceOneModel<Document> operation = new ReplaceOneModel<>(
                    eq("name", "Original Pizza"),
                    new Document("name", "Mongo's Pizza")
                            .append("borough", "Manhattan"));
            // end-bulk-replace-one

            // start-bulk-delete-one
            DeleteOneModel<Document> operation = new DeleteOneModel<>(
                    eq("restaurant_id", "5678"));
            // end-bulk-delete-one

            // start-bulk-delete-many
            DeleteManyModel<Document> operation = new DeleteManyModel<>(
                    eq("name", "Mongo's Deli"));
            // end-bulk-delete-many

            // start-bulk-write-mixed
            Publisher<BulkWriteResult> bulkWritePublisher = restaurants.bulkWrite(
                    Arrays.asList(new InsertOneModel<>(
                                    new Document("name", "Mongo's Deli")
                                            .append("cuisine", "Sandwiches")
                                            .append("borough", "Manhattan")
                                            .append("restaurant_id", "1234")),
                            new InsertOneModel<>(new Document("name", "Mongo's Deli")
                                            .append("cuisine", "Sandwiches")
                                            .append("borough", "Brooklyn")
                                            .append("restaurant_id", "5678")),
                            new UpdateManyModel<>(eq("name", "Mongo's Deli"),
                                    set("cuisine", "Sandwiches and Salads")),
                            new DeleteOneModel<>(eq("restaurant_id", "1234"))));

            BulkWriteResult bulkResult = Mono.from(bulkWritePublisher).block();

            System.out.println(bulkResult.toString());
            // end-bulk-write-mixed

            // start-bulk-write-unordered
            Publisher<BulkWriteResult> bulkWritePublisher = restaurants.bulkWrite(
                    Arrays.asList(new InsertOneModel<>(
                                    new Document("name", "Mongo's Deli")
                                            .append("cuisine", "Sandwiches")
                                            .append("borough", "Manhattan")
                                            .append("restaurant_id", "1234")),
                            new InsertOneModel<>(new Document("name", "Mongo's Deli")
                                            .append("cuisine", "Sandwiches")
                                            .append("borough", "Brooklyn")
                                            .append("restaurant_id", "5678")),
                            new UpdateManyModel<>(eq("name", "Mongo's Deli"),
                                    set("cuisine", "Sandwiches and Salads")),
                            new DeleteOneModel<>(eq("restaurant_id", "1234"))), 
                            new BulkWriteOptions().ordered(false));

            BulkWriteResult bulkResult = Mono.from(bulkWritePublisher).block();

            System.out.println(bulkResult.toString());
            // end-bulk-write-unordered

            }
    }

}

File: source/includes/write/bulk-write.py
# start-bulk-insert-one
operation = InsertOne(
    namespace="sample_restaurants.restaurants",
    document={
        "name": "Mongo's Deli",
        "cuisine": "Sandwiches",
        "borough": "Manhattan",
        "restaurant_id": "1234"
    }
)
# end-bulk-insert-one

# start-bulk-insert-one-typed
class Restaurant (TypedDict):
    name: str
    cuisine: str
    borough: str
    restaurant_id: str

operation = pymongo.InsertOne(Restaurant(
    name="Mongo's Deli", cuisine="Sandwiches", borough="Manhattan", restaurant_id="1234"))
# end-bulk-insert-one-typed

# start-bulk-update-one
operation = UpdateOne(
    namespace="sample_restaurants.restaurants",
    filter={ "name": "Mongo's Deli" },
    update={ "$set": { "cuisine": "Sandwiches and Salads" }}
)
# end-bulk-update-one

# start-bulk-update-many
operation = UpdateMany(
    namespace="sample_restaurants.restaurants",
    filter={ "name": "Mongo's Deli" },
    update={ "$set": { "cuisine": "Sandwiches and Salads" }}
)
# end-bulk-update-many

# start-bulk-replace-one
operation = ReplaceOne(
    namespace="sample_restaurants.restaurants",
    filter={ "restaurant_id": "1234" },
    replacement={
        "name": "Mongo's Pizza",
        "cuisine": "Pizza",
        "borough": "Brooklyn",
        "restaurant_id": "5678"
    }
)
# end-bulk-replace-one

# start-bulk-replace-one-typed
class Restaurant (TypedDict):
    name: str
    cuisine: str
    borough: str
    restaurant_id: str

operation = pymongo.ReplaceOne(
    { "restaurant_id": "1234" },
    Restaurant(name="Mongo's Pizza", cuisine="Pizza", borough="Brooklyn", restaurant_id="5678")
)
# end-bulk-replace-one-typed

# start-bulk-delete-one
operation = DeleteOne(
    namespace="sample_restaurants.restaurants",
    filter={ "restaurant_id": "5678" }
)
# end-bulk-delete-one

# start-bulk-delete-many
operation = DeleteMany(
    namespace="sample_restaurants.restaurants",
    filter={ "name": "Mongo's Deli" }
)
# end-bulk-delete-many

# start-bulk-write-mixed-collection
operations = [
    InsertOne(
        document={
            "name": "Mongo's Deli",
            "cuisine": "Sandwiches",
            "borough": "Manhattan",
            "restaurant_id": "1234"
        }
    ),
    InsertOne(
        document={
            "name": "Mongo's Deli",
            "cuisine": "Sandwiches",
            "borough": "Brooklyn",
            "restaurant_id": "5678"
        }
    ),
    UpdateMany(
        filter={ "name": "Mongo's Deli" },
        update={ "$set": { "cuisine": "Sandwiches and Salads" }}
    ),
    DeleteOne(
        filter={ "restaurant_id": "1234" }
    )
]

results = restaurants.bulk_write(operations)

print(results)
# end-bulk-write-mixed-collection


# start-bulk-write-mixed-client
operations = [
    InsertOne(
        namespace="sample_mflix.movies",
        document={
            "title": "Minari",
            "runtime": 217,
            "genres": ["Drama", "Comedy"]
        }
    ),
    UpdateOne(
        namespace="sample_mflix.movies",
        filter={ "title": "Minari" },
        update={ "$set": { "runtime": 117 }}
    ),
    DeleteMany(
        namespace="sample_restaurants.restaurants",
        filter={ "cuisine": "French" }
    )
]

results = client.bulk_write(operations)

print(results)
# end-bulk-write-mixed-client

# start-bulk-write-unordered
results = restaurants.bulk_write(operations, ordered=False)
# end-bulk-write-unordered

# start-bulk-write-verbose
results = client.bulk_write(operations, verbose_results=True)
# end-bulk-write-verbose